### DISCUSSION TITLE
Upgrade JGroups and support interoperability

### DISCUSSION THREAD

**Discussion by nandini1511** [belaban/JGroups#822] [Q&A]
Is there more comprehensive documentation on the jgroups-upgrade project compared to what is present in the README section?
I'm looking to upgrade JGroups from 3.3.5 to a more recent version 5.3.8 but will need to support nodes that are running on the old version.
What will need to be done on the nodes running old Jgroups version for them to form a cluster on the new nodes.

**Comment by belaban** [comment]
[1] describes the process in detail, have you taken a look? However, the disadvantage of upgrade-jgroups is that it would need to be back-ported to JGroups 3.x... This should be trivial though... If you an already back-ported project, RollingUpgrades [2] might be the better choice, as it already contains UPGRADE protocols for 3.x, 4.x and 5.x.

The UpgradeServer is always the same code; it needs to be started before the upgrade. Then, members of the old cluster need to be restarted with `UPGRADE` in their config, one by one.

Then the new 5.x servers need to be started with `UPGRADE` in their stack, too. After the 3.x nodes are gone, the new 5.x nodes can be restarted one by one with `UPGRADE` removed from their stack, or just made inactive.
Hope this helps,

[1] https://github.com/jgroups-extras/upgrade-common
[2] https://github.com/jgroups-extras/RollingUpgrades

**Reply by nandini1511** [reply]
Hi Bela,

Thank you for your response. Upon going through the suggested solutions, it seems highly unlikely that we can make a change on our existing nodes running JGroups 3.x as many of our customers are using old versions of our product that uses JGroups 3.x.
If we decide to continue using JGroups 3.x, do you see any issues, particularly because we are upgrading to RHEL8 and JDK11 from JDK1.8.

**Comment by belaban** [comment]
3.x is not supported anymore, no backports from 5.x or 4.x branches, so you're on your own... OTOH, if your setup has always worked and load doesn't increase, then you can ride 3.x into the sunset :-)

**Reply by nandini1511** [reply]
Thank you Bela....one more question. Wanted to know what has changed between version 3.3.5 and 5.3.8 that there is no interoperability between these two versions.

**Reply by belaban** [reply]
Quite a lot:
```
git diff --shortstat JGroups-3.3.5.Final jgroups-5.3.8.Final
 1235 files changed, 123447 insertions(+), 100527 deletions(-)
```

I suggest take a look at [1] ('Released') to see which issues were resolved between 3.3.5 and 5.3.8. Note that 3.3.5 was released in Aug 2013, so ~11 years ago!

[1] https://issues.redhat.com/projects/JGRP
